### PR TITLE
Add Windows Phone(WNS) support

### PR DESCRIPTION
**Pull request by krim**

### CODE REVIEW

File: lib/ruby-push-notifications/wns/wns_connection.rb
Comment (calonso) [inline lib/ruby-push-notifications/wns/wns_connection.rb]: I think we're missing an error path here. What if the connection to `login.live.com`.... fails? (i.e: returns a different error code such as 401, 500, etc...)

We're also missing a test case for that.

File: spec/ruby-push-notifications/wns/wns_connection_spec.rb
Comment (calonso) [inline spec/ruby-push-notifications/wns/wns_connection_spec.rb]: Instead of stubbing `WNSConnection.access_token` to raise an error I'd like to stub the appropriated http request to exercise those lines and still assert we're getting the error.

File: lib/ruby-push-notifications/wns/wns_connection.rb
Comment (calonso) [inline lib/ruby-push-notifications/wns/wns_connection.rb]: Shouldn't `https://login.live...` be a constant?

File: spec/ruby-push-notifications/wns/wns_pusher_spec.rb
Comment (calonso) [inline spec/ruby-push-notifications/wns/wns_pusher_spec.rb]: Where is this `public_send` defined? I can't find it

File: spec/ruby-push-notifications/wns/wns_pusher_spec.rb
Comment (calonso) [inline spec/ruby-push-notifications/wns/wns_pusher_spec.rb]: Can we iterate here over the actual notifications being pushed one line above? This one is less robust I think as if someone empties the contents of `WP_TARGETS`, the test will always pass regardless. Does it make sense?

File: lib/ruby-push-notifications/wns/wns_connection.rb
Comment (calonso) [inline lib/ruby-push-notifications/wns/wns_connection.rb]: Should this and `scope` below be constants as well?

File: lib/ruby-push-notifications/wns/wns_connection.rb
Comment (calonso) [inline lib/ruby-push-notifications/wns/wns_connection.rb]: Rather than extracting, this method is actually capitalising the headers. `capitalize_headers` would be a better name I guess.

File: spec/ruby-push-notifications/wns/wns_notification_spec.rb
Comment (calonso) [inline spec/ruby-push-notifications/wns/wns_notification_spec.rb]: I'd like to have a bit more complicated test for this case to have the recursive `build_hash` covered

File: spec/ruby-push-notifications/wns/wns_notification_spec.rb
Comment (calonso) [inline spec/ruby-push-notifications/wns/wns_notification_spec.rb]: I actually think this one and the following one are very important. What happens if `xml.text!` receives `nil`?

File: spec/ruby-push-notifications/wns/wns_notification_spec.rb
Comment (calonso) [inline spec/ruby-push-notifications/wns/wns_notification_spec.rb]: I'd like to see this test case extended a little bit to cover `launch_params` as well

File: spec/ruby-push-notifications/wns/wns_notification_spec.rb
Comment (calonso) [inline spec/ruby-push-notifications/wns/wns_notification_spec.rb]: Very nice! 🎉

**Review by calonso — CHANGES_REQUESTED** [review]
Amazing work overall. Congratulations @krim. I just left a few minor comments and nits and we'll be ready to merge!

Congratulations and thank you very much again. This is great work!

File: spec/ruby-push-notifications/wns/wns_pusher_spec.rb
Comment (krim) [reply spec/ruby-push-notifications/wns/wns_pusher_spec.rb]: It's a ruby method: https://apidock.com/ruby/Object/public_send

File: spec/ruby-push-notifications/wns/wns_connection_spec.rb
Comment (krim) [reply spec/ruby-push-notifications/wns/wns_connection_spec.rb]: I'm stubbing http request here: https://github.com/calonso/ruby-push-notifications/pull/21/files/d9f0ded80f57c11a387d54fcb06bdb5a2a9e28c3#diff-83d90512d15d2cb39e7dbe7f8051e98aR60
And not to stub WNSConnection.access_token as you can see

File: lib/ruby-push-notifications/wns/wns_connection.rb
Comment (krim) [reply lib/ruby-push-notifications/wns/wns_connection.rb]: fixed, i moved getting access token to other class

File: lib/ruby-push-notifications/wns/wns_connection.rb
Comment (krim) [reply lib/ruby-push-notifications/wns/wns_connection.rb]: fixed

File: lib/ruby-push-notifications/wns/wns_connection.rb
Comment (krim) [reply lib/ruby-push-notifications/wns/wns_connection.rb]: fixed. i'm remove raise exception and now i build correct response

File: lib/ruby-push-notifications/wns/wns_connection.rb
Comment (krim) [reply lib/ruby-push-notifications/wns/wns_connection.rb]: you are right, it will be better naming

File: spec/ruby-push-notifications/wns/wns_connection_spec.rb
Comment (krim) [reply spec/ruby-push-notifications/wns/wns_connection_spec.rb]: i rewrite getting access token from scratch and move it to separate class: https://github.com/calonso/ruby-push-notifications/pull/21/files#diff-e0db94d3ce0c02f743618ce39045ca6eR1

File: spec/ruby-push-notifications/wns/wns_pusher_spec.rb
Comment (krim) [reply spec/ruby-push-notifications/wns/wns_pusher_spec.rb]: I'm setted push types in spec manualy. Now if anyone remove types from WP_TARGETS - it wont' be passed

File: spec/ruby-push-notifications/wns/wns_notification_spec.rb
Comment (krim) [reply spec/ruby-push-notifications/wns/wns_notification_spec.rb]: added

File: spec/ruby-push-notifications/wns/wns_notification_spec.rb
Comment (krim) [reply spec/ruby-push-notifications/wns/wns_notification_spec.rb]: added

**Comment by calonso** [comment]
That's amazing @krim. Great work!! 
Just one bit left to go, are you up for adding the `WNSNotification` `it 'validates the wns format'` and `it 'validates the data'` specs on this PR or would you rather do it afterwards?

I'd rather have them added now as you have all the things fresh, but I'd leave it up to you. I'm already very happy with the work you put here.

Congratulations and thanks again!

**Comment by krim** [comment]
@calonso i add more spec for the most cases to wns_notification.
I think now we have full coverage for it.
So, we don't need anymore specs `it 'validates the wns format'` and `it 'validates the data'`. Specs for all push types covered this.

**Comment by calonso** [comment]
Great work!! One last question, what happens with those "empty" notifications when pushed to WNS? Are they accepted? delivered? rejected?

Just wondering if it would be better to reject the notification with an error instead of building an empty one. It all depends on what WNS does with them.

**Comment by krim** [comment]
@calonso WNS accept push with empty data. The push on phone shows as(toast example):
"AppName"(For example: Facebook)
"Новое уведомление"(In russian). In English i think you can see "New notification"

Do you want to reject empty pushes?  In this case you can send push with only title or only message. It shows on phone correctly. I think the person, who will use this gem must validate empty data before send pushes.

**Review by calonso — APPROVED** [review]

**Comment by calonso** [comment]
Cool, I just wanted to make sure that we were not pushing notifications that could be rejected by WNS, if empty ones are a valid use case then I'm fine with it and you're right, the person using the gem should decide wether it makes sense or not for them to send empty notifications or not.

I'm approving and merging the PR, but before we need the Changelog updated and the gem version bumped. Can you please add those two final bits before merging and releasing a new version?

**Comment by krim** [comment]
@calonso done 👍

**Pull request merged**
